Python/pythonrun.c

```diff
@@ -1485,6 +1485,70 @@ PyOS_CheckStack(void)
 
 #endif /* WIN32 && _MSC_VER */
 
+#if defined(__APPLE__)
+
+#define MAC_OS_10_9 13
+#define MAC_OS_10_10 14
+#define MAC_OS_10_11 15
+
+#include <sys/utsname.h>
+
+/*
+ * Return non-zero when we run out of memory on the stack; zero otherwise.
+ */
+
+__thread size_t last_remains = 0; /* Size of the last stack remain space */ 
+__thread size_t stack_space = 0; /* Size of a thread stack space */
+
+int
+PyOS_CheckStack(void)
+{
+    const pthread_t p_thread = pthread_self();
+    const uintptr_t end = (uintptr_t)pthread_get_stackaddr_np(p_thread);
+    const uintptr_t frame = (uintptr_t)__builtin_frame_address(0);
+
+    if (stack_space == 0) {
+        if (pthread_main_np() == 1) {
+            // On macOS 10.09 - 10.11 pthread_get_stacksize_np 
+            // returns wrong stack size on main thread.
+            // ref: https://github.com/rust-lang/rust/issues/43347#issuecomment-316783599
+            // ref: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8020753
```

Comment:
Sigh... I guess we have little choice here. I don't have access to a system running a problematic version of macOS at the moment and will verify this code later (I do have VMs running older macOS versions at home)